code/names.sql

```diff
@@ -0,0 +1,71 @@
+-- Output file
+-- Búa til töflu til að geyma nafna fjöldann
+CREATE TABLE IF NOT EXISTS names (
+    name TEXT NOT NULL,
+    year INTEGER NOT NULL,
+    frequency INTEGER NOT NULL,
+    type TEXT,
+    PRIMARY KEY (name, year, type)
+);
+
+-- Setja gögnin í töflu
+-- Merkja að ég sé að lesa inn gögn frá csv skrám
+.mode csv 
+-- importa gögnin
+.import 'data\first_names_freq.csv' names
+-- Setja öll gögn fyrir "first_name" sem first_names
+UPDATE names SET type = 'first_name' WHERE type IS NULL; 
+
+-- Gera það sama við seinni skránni nema middle_name
+.import 'data\middle_names_freq.csv' names
+UPDATE names SET type = 'middle_name' WHERE type IS NULL;
+
+-- Output file sem texta skrá
+.output names_output.txt
+
+-- Svör á spurningum:
+
+-- Spurning 1
+-- Prenta út spurninguna 
+SELECT 'Spurning 1: Hvaða hópmeðlimur á algengasta eiginnafnið??' AS question;
+-- Setja niðurstöðurnar sem summa af öllu count eftir nöfnunum og prenta út hvaða nafn hefur flest count s.s. er alngengast
+SELECT name, SUM(frequency) AS total_frequency
+FROM names
+-- Leita af eignarnöfnunum okkar þrem og prenta út svarið með fjöldanum af count fyrir það nafn sem er alngengast
+WHERE type = 'first_name' AND name IN ('Þráinn', 'Ásdís', 'Gígja')
+GROUP BY name
+ORDER BY total_frequency DESC
+LIMIT 1;
+
+-- Spurning 2
+-- Prenta út spurninguna
+SELECT 'Spurning 2: Hvenær voru öll nöfnin vinsælust?' AS question;
+-- Þessi kóði les bæði fyrstu og milli nöfnin okkar og skilar þeirri línu (row) sem hefur flest count s.s. hvenær hvert nafn var vinsælast
+SELECT name, year
+FROM (
+    SELECT name, year, frequency,
+    ROW_NUMBER() OVER (PARTITION BY name ORDER BY frequency DESC) AS row_num
+    FROM names
+    WHERE (type = 'first_name' AND name IN ('Ásdís', 'Gígja', 'Þráinn'))
+    UNION ALL
+    SELECT name, year, frequency,
+    ROW_NUMBER() OVER (PARTITION BY name ORDER BY frequency DESC) AS row_num
+    FROM names
+    WHERE (type = 'middle_name' AND name IN ('Marín', 'Ágúst'))
+) AS subquery
+WHERE row_num = 1;
```

Comment:
Smá overkill. Þið geti sleppt undirfyrirspurn bara `group by` og `having`:

```suggestion
SELECT name, year AS most_popular_year
FROM names
WHERE (name IN ('Ásdís', 'Gígja', 'Þráinn') AND type = 'first_name')
   OR (name IN ('Marín','Ágúst') AND type = 'middle_name')
GROUP BY name, type -- group by name and type
HAVING cnt = MAX(cnt) -- and limit the results to the year with the highest frequency
ORDER BY 1, 2;
```